File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/FavoritePlaylist.cs
namespace IteratorDesignPattern.Example3;

public class FavoritePlaylist : IAggregate<Song>
{
    private readonly HashSet<Song> _favorites = new();

    public bool AddFavorite(Song song)
    {
        return _favorites.Add(song);
    }

    public bool RemoveFavorite(Song song)
    {
        return _favorites.Remove(song);
    }

    public HashSet<Song> GetFavorites() => _favorites;

    public IIterator<Song> CreateIterator()
    {
        return new FavoriteIterator(this);
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/src/main/java/behavioral/iterator/example3/FavoritePlaylist.java
package behavioral.iterator.example3;

import java.util.*;

public class FavoritePlaylist implements IAggregate<Song> {
    private final Set<Song> favorites = new LinkedHashSet<>();

    public boolean addFavorite(Song song) { return favorites.add(song); }
    public boolean removeFavorite(Song song) { return favorites.remove(song); }
    public Set<Song> getFavorites() { return favorites; }

    @Override
    public IIterator<Song> createIterator() { return new FavoriteIterator(this); }
}
